File: src/safety/verdict.py
"""Verdict logic: turn a prediction (or an OOD reject) into a conservative, never-say-safe call.

Pure functions (no I/O, no model) — the most important thing to unit-test, since the BRD safety
rules (§10.1) live here: never declare safe, default to danger when unsure.
"""
from __future__ import annotations

from config import HIGH_CONFIDENCE
from schemas import DangerLevel, Prediction, Verdict


def verdict_for_prediction(pred: Prediction) -> Verdict:
    """Map calibrated P(venomous) to an honesty-graded verdict.

    Three bands: likely venomous (>=50%), probably-not-but-unconfirmed (threshold..50%), and
    likely non-venomous (<threshold). Crucially, NONE says "safe to approach" — the middle/low
    bands still forbid handling and insist on medical care if bitten (BRD never-say-safe).
    """
    t, p = pred.threshold, pred.venom_probability
    if p >= HIGH_CONFIDENCE:
        return Verdict(
            DangerLevel.DANGEROUS, "☠️", "Treat as DANGEROUS — likely venomous",
            "This is most likely a venomous or dangerous snake. Keep well back and do not approach.",
            treat_as_dangerous=True, venom_probability=p,
        )
    if p >= t:
        return Verdict(
            DangerLevel.CAUTION, "⚠️", "Best treated with caution — keep clear",
            "Some signs of risk, but no confident venom match. Stay back, don't corner or handle it, "
            "and seek medical help immediately if bitten.",
            treat_as_dangerous=True, venom_probability=p,
        )
    return Verdict(
        DangerLevel.LOW_RISK, "🟢", "Likely non-venomous — still keep your distance",
        "Low chance of venom, but never handle or approach any snake. If bitten, seek medical care immediately.",
        treat_as_dangerous=False, venom_probability=p,
    )


def verdict_for_ood() -> Verdict:
    """OOD reject: don't guess a species; tell the user to re-shoot and default to dangerous (FR-02)."""
    return Verdict(
        DangerLevel.UNIDENTIFIED, "❓", "Couldn't confirm a snake — assume DANGEROUS",
        "This doesn't look like a clear, in-scope snake photo. Re-shoot from a safe distance "
        "(zoom in, don't approach). When in doubt, stay back and treat it as dangerous.",
        treat_as_dangerous=True, venom_probability=None,
    )

File: mobile/src/safety/verdict.ts
// Honesty-graded verdict logic — ported 1:1 from Python `safety/verdict.py`.
// The safety rules live here: never declare "safe", default to danger when unsure.
// Copy comes from the active i18n dictionary (English canonical in i18n/strings/en.ts).
import { s } from "../i18n";
import { DangerLevel, Prediction, Verdict } from "./types";

// >= this venom prob -> "likely venomous" (red); between the model threshold and this -> "keep clear" (amber).
export const HIGH_CONFIDENCE = 0.5;

export function verdictForPrediction(pred: Prediction): Verdict {
  const { threshold: t, venomProbability: p } = pred;
  const v = s().verdict;
  if (p >= HIGH_CONFIDENCE) {
    return {
      level: DangerLevel.DANGEROUS,
      icon: "☠️",
      headline: v.dangerous.headline,
      subtext: v.dangerous.subtext,
      treatAsDangerous: true,
      venomProbability: p,
    };
  }
  if (p >= t) {
    return {
      level: DangerLevel.CAUTION,
      icon: "⚠️",
      headline: v.caution.headline,
      subtext: v.caution.subtext,
      treatAsDangerous: true,
      venomProbability: p,
    };
  }
  return {
    level: DangerLevel.LOW_RISK,
    icon: "🟢",
    headline: v.lowRisk.headline,
    subtext: v.lowRisk.subtext,
    treatAsDangerous: false,
    venomProbability: p,
  };
}

export function verdictForOod(): Verdict {
  const v = s().verdict;
  return {
    level: DangerLevel.UNIDENTIFIED,
    icon: "❓",
    headline: v.unidentified.headline,
    subtext: v.unidentified.subtext,
    treatAsDangerous: true,
    venomProbability: null,
  };
}
